source/matplot/axes_objects/vectors.cpp

```diff
@@ -97,13 +142,24 @@ namespace matplot {
                     ss << "  " << 0.0;
                 }
 
+                if (normalize_) {
+                    double mag = std::sqrt(u_value * u_value + v_value * v_value +
+                                      w_value * w_value);
+                    u_value *= scale_ / mag;
+                    v_value *= scale_ / mag;
+                    w_value *= scale_ / mag;
```

Comment:
To prevent division by zero in the event u_value, v_value, and w_value are are zero (then vector points in no direction).
```suggestion
                   if (mag != 0.0) {
                        u_value *= scale_ / mag;
                        v_value *= scale_ / mag;
                        w_value *= scale_ / mag;
                    }
```

Reply:
Isn't the scale usually multiplied by the magnitude? I mean, shouldn't a larger magnitude make things larger regardless of the other normalization scale?

Reply:
Sorry. I just noticed this is inside an `if (normalize_)`. This is still seems weird though. If the ultimate effect of this normalization is just to attribute a constant to these values, isn't is easier just attribute this constant directly?

Reply:
This seems like an edge case anyway, but maybe the ultimate intention is a unit_vector times `scale_` is the direction of `<u,v,w>` regardless of mag?

Reply:
Yes it is supposed to be unit_vector*`scale_` in the direction of <u,v,w>. In the `axes_type::quiver` or `axes_type::quiver3` functions in the core/axes_type.cpp file, if scale is not 0, then u, v, and w are rescaled already:

    u_scaled = (u/u_max) * scale * x_max 

when we take the magnitude of <u_scaled,v_scaled,w_scaled> (which is <u,v,w> in the vectors.cpp), the scale gets implicitly pulled out and cancels with the scale above, and then all vectors go to unit length (length of 1). To adjust back to the scale that the user specified, we need to multiple again by scale. 

We could instead have normalize_ be a float (or double) instead of bool so that it would be

    float normalize_{0};
    float vectors::normalize() const { return normalize_; }
    class vectors &vectors::normalize(float normalize) {
        normalize_ = normalize;
        touch();
        return *this;
    }

Then in the plot_data() function it would multiple u,v,w by `*= normalize_/mag` instead.

Reply:
There is also the issue that I just saw that if the user specifies `scale = 0`, then all the vectors will go to length 0 if normalized, when they should go to unit length.

Reply:
I forgot about the normalization in `axes_type::quiver` and `axes_type::quiver3`. It seems like a `float normalize_` would be redundant/ambiguous with `scale_`.

I'm worried the normalization in `axes_type::quiver` and `axes_type::quiver3` could make the current `vectors::normalize_` redundant or ambiguous? If so, it seems like a lazy normalization would be better for you. Instead of calculating unit vectors in `axes_type::quiver` and `axes_type::quiver3`, we could just set normalize_ to `true` in there, and you would later be free normalize it however you like.

In any case, this is an edge case and even any redundancy is hidden from the public API, so it's really up to you. I'm also fine with resolving this as it is if you'd like.

Reply:
> There is also the issue that I just saw that if the user specifies scale = 0, then all the vectors will go to length 0 if normalized, when they should go to unit length.

Beyond making this normalization / scales lazy, maybe it seems like a good idea to rename these things. For instance, there are many ways to "normalize" things. Maybe something like `bool unit_vectors_` would make what's happening clearer (unit vector * scale if true or direction * scale if false).

Also, `magnitudes` could just be `colors` or something like that, since this is an independent variable. A "magnitude" might make it seems like it's being inferred as `z=std::sqrt(std::pow(a,2)+std::pow(b,2)+...)` with zero memory cost instead of getting the values from another vector, which would provide the colors even if these don't match the vector magnitudes.

Reply:
Ah, but they are not normalized in `axes_type::quiver` and `axes_type::quiver3` but just rescaled. So `vectors::normalize_` is not redundant (unless the user gives normalized vectors to begin with). 

Let me make the commit for making sure that if scale = 0 that we only divide by the mag instead since the user does not want to scale and only normalize vectors.

Reply:
I think part of the confusion is whether `*= scale_` should come before or after `normalization_` to unit vectors.

Reply:
> Beyond making this normalization / scales lazy, maybe it seems like a good idea to rename these things. For instance, there are many ways to "normalize" things. Maybe something like `bool unit_vectors_` would make what's happening clearer (unit vector * scale if true or direction * scale if false).

So change `normalize_` to `unit_vectors_` and `normalize(bool)` to `unit_vectors(bool)`. Sounds good. 

> Also, `magnitudes` could just be `colors` or something like that, since this is an independent variable. A "magnitude" might make it seems like it's being inferred as `z=std::sqrt(std::pow(a,2)+std::pow(b,2)+...)` with zero memory cost instead of getting the values from another vector, which would provide the colors even if these don't match the vector magnitudes.

Sure, I will make that change. I will change magnitudes->colors and `m_data_` to `c_data_`.
 
> Vectors of length zero when scale = 0 seem like expected behavior.
It is only that in the `axes_type::quiver` and `axes_type::quiver3` if `scale = 0`, then you do not rescale the vectors at all, so it would be consistence with that.

Reply:
Great. I'm brainstorming here. Feel free to discard any of these suggestions.

Reply:
> I think part of the confusion is whether `*= scale_` should come before or after `normalization_` to unit vectors.
> 
> It's probably more useful to have `scale_` applied after normalization so both can be used.

I am following the matplotlib convention here I guess. In python I would do

    q = ax.quiver(x, y, z, u, v, w, colors=c, length=0.25, normalize=True)

for a 3d vector field. So it sets the length of each vector to the same length, and then scales them so that each vector is only 0.25 units long.  So scale should occur after normalize otherwise there is no need to specify a scale parameter.

Reply:
> So scale should occur after normalize otherwise there is no need to specify a scale parameter

Exactly. Our problem is the scaling in `axes_type` is forcing the scaling to come before normalization.

> I am following the matplotlib convention here I guess. 

This function definition looks like a good API. 

- Your `magnitudes` become `colors` because they don't really need to match the magnitudes and the magnitudes are not coming from <u,v,w>. 
- The name `length` is also clearer than `scale`, because it make it clear that it will be applied even when normalization is `true`. The problem is this would need to happen lazily in `vectors` so that scaling could come after normalization. Or `normalize` will need to go inside `axes_type`, but this doesn't seems like a good idea because it wouldn't be able be undone later like the way you did initially.
- `normalize` and `length` are not ambiguous, which is nice. However, `length` would make it seem like it's *only* being applied when `normalize` is `true`, which is not currently true for matplotplusplus, or our API would break. So `scale` is probably the best name for us.
- `normalize` is also a reasonable name in this case, because it's not really going to be a unit vector after we change its length. I don't know if there's a better name for that.

Reply:
Okay I changed the `m_data_` and `std::vector<double> m` to `c_data_` and `std::vector<double> c` where appropriate, and changed comments to exclude magnitude and use color or color mapping. 

I agree with not putting normalize in the `axes_type` since it cannot be undone. It might be best then to place the scaling in the vectors class rather than in `axes_type::quiver`. Then the raw u,v,w data is stored in the class. Upon data_string() call we can then calculate u_scaled, v_scaled, w_scaled, or if normalize_ is true, scale the resulting vector by that amount. A possible vector constructor would be

    vectors(class axes_type *parent, const std::vector<double> &x_data,
                  const std::vector<double> &y_data,
                  const std::vector<double> &z_data,
                  const std::vector<double> &u_data,
                  const std::vector<double> &v_data,
                  const std::vector<double> &w_data,
                  const std::vector<double> &c_data,
                  double scale = 1.0, std::string_view line_spec = "");

In the `vectors::data_string()` method I would add something similar to (which is what is currently used)

         auto x_copy = x;
         std::sort(x_copy.begin(), x_copy.end());
         x_copy.resize(std::distance(x_copy.begin(),
                                     std::unique(x_copy.begin(), x_copy.end())));
         std::vector<double> x_diff(x_copy.size());
         std::adjacent_difference(x_copy.begin(), x_copy.end(), x_diff.begin());
         auto xdiffmin = std::min_element(x_diff.begin() + 1, x_diff.end());
         double u_max = *std::max_element(u.begin(), u.end());
          double x_max = xdiffmin != x_diff.end() ? *xdiffmin : u_max;

         auto y_copy = y;
         std::sort(y_copy.begin(), y_copy.end());
         y_copy.resize(std::distance(y_copy.begin(),
                                     std::unique(y_copy.begin(), y_copy.end())));
         std::vector<double> y_diff(y_copy.size());
         std::adjacent_difference(y_copy.begin(), y_copy.end(), y_diff.begin());
         auto ydiffmin = std::min_element(y_diff.begin() + 1, y_diff.end());
         double v_max = *std::max_element(v.begin(), v.end());
         double y_max = ydiffmin != y_diff.end() ? *ydiffmin : v_max;

         auto u_scaled =
             (scale != 0.)
                 ? transform(
                       u, [&](double u) { return (u / u_max) * scale * x_max; })
                 : u;
         auto v_scaled =
             (scale != 0.)
                 ? transform(
                       v, [&](double v) { return (v / v_max) * scale * y_max; })
                 : v;

So, if normalize_ is true the above code is not run, but instead the simple method of dividing by mag and mutliplying by scale is used. If normalize_ is false, then the above code is used instead. This also reduces code duplicity between axes_type::quiver and axes_type::quiver3

Reply:
We could make then the member variable scale_ of `vectors` an std::optional<double> so that if the user set scale_ it evaluates to true, and otherwise evaluates to false.

Reply:
> It might be best then to place the scaling in the vectors class rather than in axes_type::quiver

That sounds great.

> A possible vector constructor would be

That looks good.

> So, if normalize_ is true the above code is not run, but instead the simple method of dividing by mag and mutliplying by scale is used

Sure. When `normalize` is false, we just change the lengths according to `scale`. When `normalize` is true, we do the same, but normalize before. We multiply by `scale` in both cases, like in our discussion about the python API. 

Regarding the normalization that comes before scaling, when `normalize` is `true`. If by `mag` you now mean the new `c_`, I don't think we need to divide the scales by the colors. Seems like bad semantics. If by `mag_` you mean the magnitude from <u,v,w>, i.e. making it a unit vector before scaling, then sure. I think you meant the later but just to be sure.

> We could make then the member variable scale_ of vectors an std::optional so that if the user set scale_ it evaluates to true, and otherwise evaluates to false

We can do that, but I think we don't really need to. This might require some changes in the API while `scale_ == 1.0` can achieve the same effect. If we want to be very pedantic about the processing cost, we could have a kind of [`if (!almost_equal(scale_, 1.0))`](https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon) before scaling, but I wouldn't worry much about that either. It seems like premature optimization all things considered.